--- /home/misterdulister/Coding/buildroot_kscf52r2/output/build/linux-5.4.45/drivers/net/ethernet/freescale/fec_main.c	2020-06-07 13:18:52.000000000 +0200
+++ /home/misterdulister/Coding/buildroot_kscf52r2/changed_files/linux-3.x/drivers/net/ethernet/freescale/fec_main.c	2020-12-15 17:04:15.567695034 +0100
@@ -237,7 +237,7 @@
  * account when setting it.
  */
 #if defined(CONFIG_M523x) || defined(CONFIG_M527x) || defined(CONFIG_M528x) || \
-    defined(CONFIG_M520x) || defined(CONFIG_M532x) || defined(CONFIG_ARM) || \
+    defined(CONFIG_KSCF52xx) || defined(CONFIG_M520x) || defined(CONFIG_M532x) || defined(CONFIG_ARM) || \
     defined(CONFIG_ARM64)
 #define	OPT_FRAME_SIZE	(PKT_MAXBUF_SIZE << 16)
 #else
@@ -2200,7 +2200,7 @@
 
 /* List of registers that can be safety be read to dump them with ethtool */
 #if defined(CONFIG_M523x) || defined(CONFIG_M527x) || defined(CONFIG_M528x) || \
-	defined(CONFIG_M520x) || defined(CONFIG_M532x) || defined(CONFIG_ARM) || \
+	 defined(CONFIG_KSCF52xx) || defined(CONFIG_M520x) || defined(CONFIG_M532x) || defined(CONFIG_ARM) || \
 	defined(CONFIG_ARM64) || defined(CONFIG_COMPILE_TEST)
 static __u32 fec_enet_register_version = 2;
 static u32 fec_enet_register_offset[] = {
