--- /home/misterdulister/Coding/buildroot_kscf52r2/output/build/linux-5.4.45/drivers/net/ethernet/freescale/fec.h	2020-06-07 13:18:52.000000000 +0200
+++ /home/misterdulister/Coding/buildroot_kscf52r2/changed_files/linux-3.x/drivers/net/ethernet/freescale/fec.h	2021-01-26 11:05:36.036233618 +0100
@@ -19,7 +19,7 @@
 #include <linux/ptp_clock_kernel.h>
 #include <linux/timecounter.h>
 
-#if defined(CONFIG_M523x) || defined(CONFIG_M527x) || defined(CONFIG_M528x) || \
+#if defined(CONFIG_KSCF52xx) || defined(CONFIG_M523x) || defined(CONFIG_M527x) || defined(CONFIG_M528x) || \
     defined(CONFIG_M520x) || defined(CONFIG_M532x) || defined(CONFIG_ARM) || \
     defined(CONFIG_ARM64) || defined(CONFIG_COMPILE_TEST)
 /*
