--- /home/misterdulister/Coding/buildroot_kscf52r2/output/build/linux-5.4.45/arch/m68k/include/asm/mcfsim.h	2020-06-07 13:18:52.000000000 +0200
+++ /home/misterdulister/Coding/buildroot_kscf52r2/changed_files/linux-3.x/arch/m68k/include/asm/mcfsim.h	2021-01-26 11:04:33.636166593 +0100
@@ -22,6 +22,8 @@
 #include <asm/mcfintc.h>
 #elif defined(CONFIG_M520x)
 #include <asm/m520xsim.h>
+#elif defined(CONFIG_KSCF52xx)
+#include <asm/kscf52xxsim.h>
 #elif defined(CONFIG_M523x)
 #include <asm/m523xsim.h>
 #include <asm/mcfintc.h>
