--- output/build/linux-5.10/arch/m68k/include/asm/mcfsim.h	2020-12-13 23:41:30.000000000 +0100
+++ changed_files/linux-3.x/arch/m68k/include/asm/mcfsim.h	2020-12-15 17:04:15.563696026 +0100
@@ -1,16 +1,15 @@
-/* SPDX-License-Identifier: GPL-2.0 */
 /****************************************************************************/
 
 /*
- *	mcfsim.h -- ColdFire System Integration Module support.
+ *  mcfsim.h -- ColdFire System Integration Module support.
  *
- *	(C) Copyright 1999-2003, Greg Ungerer (gerg@snapgear.com)
- * 	(C) Copyright 2000, Lineo Inc. (www.lineo.com)
+ *  (C) Copyright 1999-2003, Greg Ungerer (gerg@snapgear.com)
+ *  (C) Copyright 2000, Lineo Inc. (www.lineo.com)
  */
 
 /****************************************************************************/
-#ifndef	mcfsim_h
-#define	mcfsim_h
+#ifndef mcfsim_h
+#define mcfsim_h
 /****************************************************************************/
 
 /*
@@ -20,6 +19,8 @@
 #if defined(CONFIG_M5206) || defined(CONFIG_M5206e)
 #include <asm/m5206sim.h>
 #include <asm/mcfintc.h>
+#elif defined(CONFIG_KSCF52xx)
+#include <asm/kscf52xxsim.h>
 #elif defined(CONFIG_M520x)
 #include <asm/m520xsim.h>
 #elif defined(CONFIG_M523x)
@@ -49,4 +50,4 @@
 #endif
 
 /****************************************************************************/
-#endif	/* mcfsim_h */
+#endif  /* mcfsim_h */
